---
title: "Salmon-tximport-edgeR **DGE** analysis"
output: html_document
---

# Preprocessing

To get the salmon "expected counts", run the `run_salmon` script and spot check the results 

# Load some libraries

```{r load_libraries}
library(edgeR)
library(tximport)
library(rtracklayer)
```

# Start from a table of metadata

Read in the table of metadata and check that the files needed exist.

```{r read_metadata}
md <- read.table("metadata.csv", sep=";", 
                 header=TRUE, stringsAsFactors=FALSE)
stopifnot(file.exists(md$salmonfile))
#stopifnot(file.exists(md$bamfile))
md$col <- as.numeric(as.factor(md$group))
md
```

# Use *tximport* to read in the quantifications

```{r get_anno_call_tximport}
anno <- import("../data/anno/transcripts.gtf")
txgene <- unique(mcols(anno)[,c("transcript_id","gene_id")])

txi <- tximport(structure(md$salmonfile, names = md$sample), 
                type = "salmon", tx2gene=txgene)
```


# Setup edgeR container

We recognize, of course, that other reasonable options, such as sleuth/DESeq2/etc, exist.

```{r make_DGEList}
cts <- txi$counts
normMat <- txi$length
normMat <- normMat/exp(rowMeans(log(normMat)))
o <- log(calcNormFactors(cts/normMat)) + log(colSums(cts/normMat))

d <- DGEList(cts)
d$offset <- t(t(log(normMat)) + o)
d <- calcNormFactors(d)
```


# Always look at (multiple) diagnostics:  #1 multi-dimensional scaling plot

```{r diag_mds}
plotMDS(d, col=md$col)
```

# Filter

We usually apply a filter to remove lowly expressed transcripts/genes.

```{r filter}
k <- rowSums(cpm(d)>1) >= 3
dim(d)
d <- d[k,]
dim(d)
```

# Dispersion estimation

Define the design matrix (in this case, quite straightforward) and then estimate dispersions.

```{r design_dispersion}
mm <- model.matrix(~group,data=md)
d <- estimateDisp(d,mm)
```

# Diagnostic: BCV plots

We generally observe a decreasing trend here.

```{r diag_bcv}
plotBCV(d)
```

# Setup for the result output

This is a little book-keeping, so that the output table has the transcript identifiers and the full table of counts-per-million.  Here is the place to add other useful stuff, such as gene symbols, chromosome locations, etc.

```{r bookkeeping}
cps <- cpm(d)
d$genes <- data.frame(gene_id=rownames(d), 
                      round(cps,1))
```


# Formal statistical testing

Here, it is a simple two-group comparison, but more detailed comparisons can be easily made, using an appropriate combination of design matrix and contrasts/coefficients.

```{r fit_test}
f <- glmFit(d, mm)
lrt <- glmLRT(f, coef=2)
topTags(lrt)
```

# Diagnostic: Histogram of P-values

```{r pval_hist}
hist(lrt$table$PValue, 50)
```


# Diagnostic: Spot check of some DE transcripts
```{r known_diffT}
barplot( cps["ENSG00000117472",], col=md$col )
```


# Reproducibility

When you do these analyses, it is always a good idea to keep a record of the versions used

```{r sessioninfo}
sessionInfo()
```
